Load the FeatureAnnouncement modal to the order listing page

harriswong opened the pull request:
## Description
This PR makes the migration admin notice's "Confirm update" button functional.
![image](https://github.com/Automattic/woocommerce-services/assets/572862/0a5880a9-a046-4075-b14d-7171528628cb)

Upon clicking the button, the migration will install, and activate WCShipping and WCTax. 
![image](https://github.com/Automattic/woocommerce-services/assets/572862/0dc05686-99f5-425a-a1bd-a6f7ed2c6c64)

This is the same as clicking the "Update now" directly in the modal, introduced in this PR https://github.com/Automattic/woocommerce-services/pull/2745.

## Technical decision
### [main.js](https://github.com/Automattic/woocommerce-services/blob/trunk/client/main.js#L54) vs wcshipping-migration-admin-notice.js
To load `<FeatureAnnouncement>` on the order list page, we could use the existing `getRouteClass` pattern found in [main.js](https://github.com/Automattic/woocommerce-services/blob/trunk/client/main.js#L54). This way, we wouldn't need to create our own webpack's entry point. In order to use this pattern, we need to follow the `enqueue_wc_connect_script()`  code to load the javascript and style, etc. There are also a lot of variables added in here that we don't need. For the sake of the migration modal, I thought it would be simpler (and we'd have more control) to create our own [entry point](https://github.com/Automattic/woocommerce-services/pull/2756/files#diff-1fb26bc12ac780c7ad7325730ed09fc4c2c3d757c276c3dacc44bfe20faf166fR58) and render the `<FeatureAnnouncement>` component inside the PHP admin notice banner. This way, we also have more control over the modal without worrying about breaking other pages. The downside is loading the `isEligable` props which is from the redux store. It turned out it wasn't too bad because we only needed the [shipping label store](https://github.com/Automattic/woocommerce-services/pull/2756/files#diff-cafa06d18d10a101b583ffbe68a81c29f97cacbf981e3e79167354c18786efa2R21). As a result, I chose to stick with `client/wcshipping-migration-admin-notice.js` instead of following the `main.js` pattern.

### Clicking "Confirm update" to trigger React `<FeatureAnnouncement>` button action
Instead of creating our own React component for the admin notice. I continued to use the WP's one from the PHP's side. To trigger the migration inside the PHP's admin notice, I added an event listener to render the `<FeatureAnnouncement>` component with the store context. Then, the script continues and will click the "Update now" button to trigger the migration. 

### Related issue(s)
NA

### Steps to test
1. Go to `woocommerce-services/classes/class-wc-connect-service-settings-store.php` and replace the following `is_eligible_for_migration` function. This function decides if the modal should popup or not
```
		public function is_eligible_for_migration() {
			WC_Connect_Options::delete_option( 'wcshipping_migration_state' );
			$migration_state = WC_Connect_Options::get_option( 'wcshipping_migration_state' );
			//TODO: Return $migration_state check && connect-server flag
			return WC_Connect_WCST_To_WCShipping_Migration_State_Enum::COMPLETED !== $migration_state && true;
		}
```
2. Make sure you have "WooCommerce Shipping" and "WooCommerce Tax" in your plugins folder. We can't test the "install" part because they aren't part of dotorg yet.
3. Go to the order listing page.
4. Press F12 and check the Network tab, click the "Confirm update" button in the admin notice banner ![image]
(https://github.com/Automattic/woocommerce-services/assets/572862/0a5880a9-a046-4075-b14d-7171528628cb)
5. Confirm the 3 requests did go through and confirm the modal popped up and the button in the modal reads "Updating...".
![image](https://github.com/Automattic/woocommerce-services/assets/572862/ab43438a-2cd9-491d-999d-968e44f914a9)
6. Confirm you are being redirected to the plugins page with WCS&T deactivated, WCShipping and WCTax activated.

### Checklist

<!-- All testable code should have tests. -->
- [ ] unit tests

<!-- An entry in the changelog file is always required -->
- [ ] `changelog.txt` entry added
- [ ] `readme.txt` entry added

Commit:
Load a new javascript file when migration admin notice shows up

Commit:
Load redux store for admin notice

The redux store inlcudes only the label shipping's reducer and the initial states.

Commit:
semi-colon gets rendered out. Removing.

Commit:
Fix style and setup initial data from the PHP's side

harriswong commented:
### Jun 21 update
#### Done
- I added a new entrypoint to compile the feature announcement codes into its own js and css.
- Modal loads on order list page

#### TODO
- "Update now" button in the modal is broken. It's saying 404 not found on the order listing page. I think WCS&T isn't loaded on this page, so the controller isn't registered. This is gonna be a bit tricky to load the controller and all its dependency :/ 
- The modal shouldn't popup immediately on the order list page, it should only pop up after clicking the admin notice's "Confirm update".
- Clicking "Confirm update" should automatically click "Update now"

Commit:
Use wcs-client alias for webpack otherwise order list page can't find it

Commit:
Add baseURL and redirect URL through the proper enqueued scripts

Commit:
Move inline styles to the migration scss

Commit:
Clicking "Confirm update" will start the migration process

harriswong commented on client/wcshipping-migration-admin-notice.js:
```diff
@@ -0,0 +1,43 @@
+const wcstWCShippingMigrationNoticeButton = document.getElementById('wcst-wcshipping-migration-notice__click');
+
+// Clicking "Confirm update" will start the migration. This is the same as popping up the modal and clicking the "Update" button there.
+["click", "keydown"].forEach(eventName =>
```
ARIA, needs to bind both `click` and `keydown` https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#sect1

harriswong commented on client/wcshipping-migration-admin-notice.js:
```diff
@@ -0,0 +1,43 @@
+const container = document.getElementById('wcst_wcshipping_migration_admin_notice_feature_announcement');
+const args = container.dataset.args && JSON.parse( container.dataset.args ) || {};
+setNonce( args.nonce );
+setBaseURL( args.baseURL );
```
`getBaseURL` is used inside the `<FeatureAnnouncement>` component. It needs to be set here, similar to how we do `setBaseURL(...)` in `main.js`.

harriswong commented on woocommerce-services.php:
```diff
@@ -1898,22 +1900,52 @@ public function display_woo_shipping_and_woo_tax_are_active_notice() {
+			wp_register_script( 'wcst_wcshipping_migration_admin_notice', $this->wc_connect_base_url . 'woocommerce-services-wcshipping-migration-admin-notice-' . $plugin_version . '.js', array(), null );
+			wp_enqueue_script( 'wcst_wcshipping_migration_admin_notice' );
+			wp_enqueue_style( 'wcst_wcshipping_migration_admin_notice' );
+			wp_enqueue_script( 'wc_connect_admin' );
```
> wp_enqueue_script( 'wc_connect_admin' );

I enqueued this whole script because `<FeatureAnnouncement>` needs the global `global.wcsPluginData.adminPluginPath`, which is a variable in the `wc_connect_admin` script. Loading this whole script is easier.

Alternatively, I could have extracted the `global.wcsPluginData.adminPluginPath`, and loads it in the [$encoded_arguments](https://github.com/Automattic/woocommerce-services/pull/2756/files#diff-4cb5834c68b324f4fbecc437b25f62021e55ecc28db9adb93053eca0391f5cc5R1912) up there. If I do this, I would have to edit `<FeatureAnnouncement>` and change:
 `window.location = global.wcsPluginData.adminPluginPath;`
to something like:
`window.location = global.wcsPluginData.adminPluginPath || someState.adminPluginPath;`.
This doesn't seem worth it because the condition is more complicated (prone to errors). So I chose to load the whole `wc_connect_admin` script into the order list page instead.

waclawjacek replied:
Thanks for the explanation. This also seems OK performance-wise because I think `/client/main.js` won't load the specific functionality (like the label purchase modal) unless certain conditions are met, correct? 🤔

harriswong replied:
Thanks @waclawjacek! Actually, you just made me realize that `client/main.js` is the registered `wc_connect_admin` script which turns into `'woocommerce-services-*.js`. 
If I am already loading `wc_connect_admin`, maybe we didn't need its own entry point. I have 3 options.

#### Option 1.
Leave it as is because this is the migration modal only. Loading `'woocommerce-services-*.js` for `adminPluginPath` is heavy but it is also cached, so it should be faster for the individual order page. The `client/wcshipping-migration-admin-notice.js` is separated as if it was chunk-split and won't be loaded inside the individual order page, which is good.

#### Option 2.
Remove `client/wcshipping-migration-admin-notice.js` and move that into `main.js`. Use `enqueue_wc_connect_script()` pattern to create the `wcc-` container inside the admin notice in order to load this `<FeatureAnnouncement>` modal.

#### Option 3.
Don't load `wp_enqueue_script( 'wc_connect_admin' );`. Pull `adminPluginPath` and find a way to load this in the order list page, and then inject it to the `<FeatureAnnouncement>` modal.

I am opting for **Option 1** because there is a slight performance hit on the order list page, but does not impact overall total performance since `woocommerce-service-*.js` is cached and will be used in the individual order page, it will be loaded regardless. On size, it will actually be better because these codes won't be loaded on the individual order page.

Let me know your thoughts!

samnajian replied:
I'd rather go with **Option 1** as it's a migration specific change.

Commit:
Remove phpcs:ignore because it does have a version

waclawjacek commented on client/components/migration/feature-announcement.jsx:
```diff
@@ -7,7 +7,7 @@ import { useState } from '@wordpress/element';
 import { bindActionCreators } from 'redux';
 import { localize } from 'i18n-calypso';
-import { getNonce, getBaseURL } from 'api/request'; // client/api/request.js
+import { getNonce, getBaseURL } from 'wcs-client/api/request';
```
Nice.

waclawjacek commented on client/wcshipping-migration-admin-notice.js:
```diff
@@ -0,0 +1,43 @@
+		);
+
+		// Click the update button in the modal to start the migration.
+		const update_button = document.getElementsByClassName('migration__announcement-update-button');
```
Maybe it'd be more precise if we used an ID instead?

harriswong replied:
Thanks! I thought about using `ID` but I couldn't find `ID` in https://developer.wordpress.org/block-editor/reference-guides/components/button/#props. 
But! Now that you brought it up again, I wonder if `ID` is "native" so it should just work? Let me give this a try.

harriswong replied:
ID works, updated in 402998bb

waclawjacek commented on client/wcshipping-migration-admin-notice.js:
```diff
@@ -0,0 +1,43 @@
+const args = container.dataset.args && JSON.parse( container.dataset.args ) || {};
+setNonce( args.nonce );
+setBaseURL( args.baseURL );
+const ShippingLabelStore = ShippingLabel(args);
+const store = createStore(ShippingLabelStore.getReducer(), ShippingLabelStore.getInitialState());
```
What are we actually using this for? Do we need to use the reducer?

harriswong replied:
> What are we actually using this for?

The `store` is needed because `<FeatureAnnouncement>` connects to the "label store" for the [isEligableToMigrate ](https://github.com/Automattic/woocommerce-services/blob/1bce661e84bf2efbc27e9bfbdd4cb2896f1f5510/client/components/migration/feature-announcement.jsx#L197-L199)selector.
In order for this selector to work without modification, the store needs to have the `extensions > woocommerce > woocommerceServices > sites >...` path 
![image](https://github.com/Automattic/woocommerce-services/assets/572862/efe5b1e2-ca6f-4c77-b5ce-4b253584eb18)
I could have created a different store by loading a different set of initial states similar to [this](https://github.com/Automattic/woocommerce-services/blob/1bce661e84bf2efbc27e9bfbdd4cb2896f1f5510/client/apps/shipping-label/index.js#L118-L137). But I chose to load the existing store instead.

> Do we need to use the reducer?

No. The `<FeatureAnnouncement>` only consumes the data from the redux store. It doesn't dispatch any action to update the store's state. Reducer isn't needed specifically for this modal. However! [createStore()](https://redux.js.org/api/createstore) requires the first arguments to be a reducer. So I use the existing ShippingLabelStore's one. I supposed we could put an empty reducer if we don't want to use the existing one? Something like:
```js
function shippingLabelReducer(state = [], action) {
   return state;
}
```

Let me know what you think!

waclawjacek reviewed (approved):
LGTM! :shipit: Posted some questions but things work and look good. Please fix the code style before merging.

harriswong commented:
> Please fix the code style before merging.

👋 Hey @waclawjacek , can you help me pinpoint where these issues are please? I am solely relying on the phpcs precommit hooks. Do you mean the code conflict?

Commit:
Use ID as the selector for the modal update button instead.

Commit:
Merge branch 'wcs-migration' into feature/wcshipping-migration-admin-notice-confirm-update

harriswong commented:
👋 Hi @waclawjacek, I fixed the conflict and addressed the concerns! Please let me know what you think.

harriswong commented:
👋 Hey @waclawjacek, please hold off on the review. I am anticipating a big merge conflict coming in https://github.com/Automattic/woocommerce-services/pull/2758. I will fix those conflicts after it's merged and will ping you again.

Commit:
Merge branch 'wcs-migration' into feature/wcshipping-migration-admin-notice-confirm-update

Commit:
Fix merge conflict issues with the existing css and js overlaps

Commit:
Change testing 10s back to 3 * 24 * 60 * 60

harriswong commented:
Hey @waclawjacek, I fixed the merge conflict and the functionality of this PR should work. Please take another look.

~Note, the dismissal button from https://github.com/Automattic/woocommerce-services/pull/2758 doesn't seem to work anymore after the merge. I ran out of energy tonight to dig into it. I can create a separate issue/PR to address that regression.~  
**edited: This is fixed. Please check the comment below**

Commit:
Fix admin notice dimissible button and cleaned up jquery

harriswong commented:
> Note, the dismissal button from https://github.com/Automattic/woocommerce-services/pull/2758 doesn't seem to work anymore after the merge. I ran out of energy tonight to dig into it. I can create a separate issue/PR to address that regression.

Hey @waclawjacek. I fixed(4df565cb15248501999fac0890ce72d1742c3ab2) the dismissible button issue after the last merge conflict. Here is a summary of what I did:
- Removed the dependencies `jQuery` and `wp-utils` in `wcst_wcshipping_migration_admin_notice`. We will use plain javascript instead.
- Replaced the javascript cookie logic with a plain javascript cookie logic. The cookie sets a `1` and will expire in 3 days.
- The WP admin notice's "is-dismissible" button is added after the DOM is ready. I tried a few things but the dismiss button wasn't found (not loaded in DOM). So I cloned the markup of the button and added it directly to our admin notice instead. Then, I bind the 3-days-cookie event to it. This way, the button is loaded into the DOM.
- Updated the scss on the banner so that it looks the same as figma.

Commit:
After clicking dimiss, remove itself from the DOM until 3 days later

waclawjacek commented:
Thanks for the changes! I won't be able to review them as I'm headed AFK. @Automattic/escargot, could someone please give the PR a look?

samnajian commented:
> Thanks for the changes! I won't be able to review them as I'm headed AFK. @Automattic/escargot, could someone please give the PR a look?

I can review it.

samnajian commented on client/wcshipping-migration-admin-notice.js:
```diff
@@ -0,0 +1,61 @@
+const wcstMigrationNoticeDimissButton = document.querySelector('.wcst-wcshipping-migration-notice button.notice-dismiss');
+
+// Helper function to set simple name=>value cookie in days.
+const setCookie = (name, value, days) => {
```
[Optional] You could use `window.wpCookies`  to set cookies.

harriswong replied:
Thanks. I like this suggestion. Replaced it with the wp's one bd141b6c

samnajian commented on client/wcshipping-migration-admin-notice.js:
```diff
@@ -0,0 +1,61 @@
+
+	// Dismiss it for 3 days, then remove it from view.
+	wcstMigrationNoticeDimissButton.addEventListener(eventName, () => {
+		setCookie('wcst-wcshipping-migration-dismissed', 1, 3);
```
[ Optional ] what do you think about assigning `3` to a variable with a transparent name with the following benefits:
1) It won't be lost in the code, but explicitly defined in this file in a higher level.
2) A proper name explains what it is.

harriswong replied:
Updated to a constant in bd141b6c

samnajian commented on client/wcshipping-migration-admin-notice.js:
```diff
@@ -0,0 +1,61 @@
+		const wcstMigrationAdminNotice = document.querySelector('.wcst-wcshipping-migration-notice');
+		wcstMigrationAdminNotice.remove();
+	});
+});
```
Missing empty line at the end of file.

harriswong replied:
Fixed in 42b2e131

samnajian reviewed (approved):
LGTM!

- Please see the added comments and feel free to apply them before merging.
- There is a merge conflict which needs care before merging.
- It functions how the issue requires you to do, but it's a bit weird that the modal opens and then you're redirected shortly to the plugins page, it was confusing to me, it can be confusing to others as well, I'd call @MatthiasReinholz to have his input, but this can probably be another issue and this PR doesn't have to wait as it addresses what it's supposed to do.

Commit:
Merge branch 'wcs-migration' into feature/wcshipping-migration-admin-notice-confirm-update

harriswong commented:
> It functions how the issue requires you to do, but it's a bit weird that the modal opens and then you're redirected shortly to the plugins page, it was confusing to me, it can be confusing to others as well, I'd call @MatthiasReinholz to have his input, but this can probably be another issue and this PR doesn't have to wait as it addresses what it's supposed to do.

100% agree. This is also discussed via the design request as well: p1718745005310189-slack-C06FP5XT1S5

Commit:
Add EOL to the js file.

harriswong commented:
~After resolving the merge conflicts, the API calls inside `migration-runner.js` is returning `undefined` for `nonce`. I am looking into it.~

**edited: Fixed in 2fbd631d**

Commit:
Change headers to a callback to retrieve getNonce() after dependencies are loaded

Commit:
Refactor cookie function to window.wpCookies

harriswong commented:
I tested the following:
- In the order listing page, click "Confirm update", I can confirm that it's installing and activating the plugins. Then it deactivates WCS&T and redirects the user to the plugin page.
- I confirmed that clicking "dismiss" sets the cookie for 3 days
![image](https://github.com/Automattic/woocommerce-services/assets/572862/0146b0c7-f457-45dc-8bdc-ae1f278dfbdd)
- In the order page itself, clicking "Update now" also installs, and activates the plugins. Then, it deactivates WCS&T and redirects to the plugins page. 👍 

@samnajian, please take another look and please feel free to merge for me if it looks good.

samnajian reviewed (approved):
LGTM! Thanks for updates.
